pymc4/flow/executor.py

```diff
@@ -33,35 +54,315 @@ class EarlyReturn(StopIteration):
     ...
 
 
-class Executor(metaclass=abc.ABCMeta):
-    @abc.abstractmethod
-    def new_state(self, *args, **kwargs):
-        raise NotImplementedError
+class SamplingState(object):
+    __slots__ = (
+        "transformed_values",
+        "untransformed_values",
+        "observed_values",
+        "all_values",
+        "distributions",
+        "potentials",
+    )
+
+    def __init__(
+        self,
+        transformed_values: Dict[str, Any] = None,
+        untransformed_values: Dict[str, Any] = None,
+        observed_values: Dict[str, Any] = None,
+        distributions: Dict[str, abstract.Distribution] = None,
+        potentials: List[abstract.Potential] = None,
+    ):
+        # verbose __init__
+        if transformed_values is None:
+            transformed_values = dict()
+        else:
+            transformed_values = transformed_values.copy()
+        if untransformed_values is None:
+            untransformed_values = dict()
+        else:
+            untransformed_values = untransformed_values.copy()
+        if observed_values is None:
+            observed_values = dict()
+        else:
+            observed_values = observed_values.copy()
+        if distributions is None:
+            distributions = dict()
+        else:
+            distributions = distributions.copy()
+        if potentials is None:
+            potentials = list()
+        else:
+            potentials = potentials.copy()
+        self.transformed_values = transformed_values
+        self.untransformed_values = untransformed_values
+        self.observed_values = observed_values
+        self.all_values = collections.ChainMap(
+            self.untransformed_values, self.transformed_values, self.observed_values
+        )
+        self.distributions = distributions
+        self.potentials = potentials
+
+    def collect_log_prob(self):
+        return sum(
```

Comment:
I'm concerned about this part. It works for scalar values, but for vector valued log_prob this will fail. We need to access backend ops to correctly perform reduce operation.

Reply:
This comment is on oudated line. Is it still a concern of yours?

Reply:
Outdated